tests/test_steps.py

```diff
@@ -0,0 +1,123 @@
+from pytest import fail
+
+from storey import build_flow, Source, Map
+from storey.dtypes import FlowError, Event
+from storey.steps import Flatten, Sample, Assert, ForEach
+from time import sleep
+
+
+def test_assert():
+    try:
+        controller = build_flow(
+            [Source(), Assert(lambda event: event > 10, raise_on_exception=True)]
+        ).run()
+        controller.emit(1)
+        controller.await_termination()
+    except FlowError:
+        return True
+
+    fail("User defined assert not failing", False)
+
+
+def test_sample_by_count():
+    class CountUntil:
+        def __init__(self, max: int):
+            self.max = max
+            self.count = 0
+
+        def __call__(self, event):
+            self.count += 1
+            return self.count <= self.max
+
+    controller = build_flow(
+        [Source(), Assert(CountUntil(5)), Sample(5), Assert(CountUntil(1))]
+    ).run()
+
+    controller.emit(1)
+    controller.emit(1)
+    controller.emit(1)
+    controller.emit(1)
+    controller.emit(1)
+    controller.terminate()
+    controller.await_termination()
+
+
+def test_sample_by_seconds():
+    class CountUntil:
+        def __init__(self, max: int):
+            self.max = max
+            self.count = 0
+
+        def __call__(self, event):
+            self.count += 1
+            return self.count <= self.max
+
+    controller = build_flow(
+        [Source(), Assert(CountUntil(6)), Sample(rate_count=2), Assert(CountUntil(2))]
+    ).run()
+
+    controller.emit(1)
+    controller.emit(1)
+    controller.emit(1)
+
+    sleep(2)
```

Comment:
Using `sleep` is not ideal in terms of run time and reliability. See [test_batch_grouping_with_timeout](https://github.com/mlrun/storey/blob/de40cf6/tests/test_flow.py#L793) for an alternative.